AliEn-Runtime removal

adriansev opened the pull request:
Hi @ktf @TimoWilken @costing I would like to propose the removal of AliEn-Runtime recipe
which brings much confusion and package duplication in building and running of packages.
The removal was simply done by replacing instances of AliEn-Runtime with it's actual components, removing duplicates with preference to `require:` (meaning if the same package ended up being in both `requires:` and `build_requires:` after replacement, only the `requires:` one was kept).
The main goal of removal was to avoid having duplicate packages in the dependency chain __WITH DIFFERENT SOURCES__ and any other mixup of dependencies declaration.
Let me know what you think .. doing the draft until this was discussed enough

davidrohr commented:
I like this :)

TimoWilken commented:
Hi @adriansev, thanks! This looks like a good idea and would simplify things fundamentally.

I don't know why AliEn-Runtime was added in the first place, especially as a dependency for other packages -- @ktf @costing do you know? Does anything depend on it being present?

@adriansev, I've had a brief look over the proposed changes, and please be careful about keeping runtime and build requirements apart! For example, `alibuild-recipe-tools` must never be a runtime `requires`, only ever a `build_requires`. Similarly for other libraries; you seem to be moving a lot of stuff from `build_requires` to `requires` without a reason that's clear to me.

It would also be great if you could solve the linter errors and warnings indicated on the "files" tab, at least on the lines you added or changed. You can also run https://github.com/alisw/alidistlint locally to see them, as GitHub unfortunately doesn't display all of them (only a maximum of 10).

adriansev commented:
Hi @TimoWilken 

> Hi @adriansev, thanks! This looks like a good idea and would simplify things fundamentally.

i really hope so :)

> @adriansev, I've had a brief look over the proposed changes, and please be careful about keeping runtime and build requirements apart! For example, `alibuild-recipe-tools` must never be a runtime `requires`, only ever a `build_requires`. Similarly for other libraries; you seem to be moving a lot of stuff from `build_requires` to `requires` without a reason that's clear to me.

so, from http://alisw.github.io/alibuild/reference.html#the-header section on `build_requires` i got that `alibuild-generate-module won’t pull in build requirements’ modulefiles.`
so, it sounded fishy to me so i put alibuild-recipe-tools in requires just to be safe .. 
but noted, i will move it to `build_requires`

for the rest of libraries my thought was : "if it is used to compiled the package, but then
when loading package the library also is not loaded, things will crash"

so, given you comment i might understood something wrong about build_requires.
so, lets have an example: xrootd
if i put libxml2 in build_requires, when loading/enter xrootd, will the libxml2 be also loaded with all the required environment?

> It would also be great if you could solve the linter errors and warnings indicated on the "files" tab, at least on the lines you added or changed. You can also run [alisw/alidistlint](https://github.com/alisw/alidistlint) locally to see them, as GitHub unfortunately doesn't display all of them (only a maximum of 10).

oh, thanks for info! i will do this shortly

TimoWilken commented:
> so, from http://alisw.github.io/alibuild/reference.html#the-header section on `build_requires` i got that `alibuild-generate-module won’t pull in build requirements’ modulefiles.` so, it sounded fishy to me so i put alibuild-recipe-tools in requires just to be safe .. but noted, i will move it to `build_requires`

Maybe it's a bit confusingly worded, but alibuild-recipe-tools only provides the `alibuild-generate-module` script, nothing else. That script is only needed at build time, in order to generate the final modulefile, so it must be a build requires. At runtime, the final generated modulefile is available, so we don't need alibuild-recipe-tools any more.

And at runtime, build requirements are not available, so it makes sense not to put them in the modulefile.

> for the rest of libraries my thought was : "if it is used to compiled the package, but then when loading package the library also is not loaded, things will crash"

Yes, it seems like the stuff that AliEn-Runtime depends on needs to be available at runtime, so you're right to put that in `requires`, except for alibuild-recipe-tools.

However, please be careful with existing packages that don't have anything to do with AliEn-Runtime. For instance, for Python, please don't make curl a runtime requirement unless there's a reason for it -- it doesn't seem to be used at runtime as far as I can see, and things work fine with it as a build requirement. (Is it even required at build time? The script doesn't seem to reference it... It seems like we used to fetch pip using curl, maybe whoever changed that forgot to remove the dependency.)

Finally, I'm not sure that just removing AliEn-Runtime from the modulefiles is a good idea. You should probably add the things that AliEn-Runtime provided to the modulefile (like OpenSSL, zlib, libxml, etc), so that these libraries are available at runtime, like before.

Of course, not every package that depends on AliEn-Runtime necessarily needs everything that it provides -- e.g. Python might need OpenSSL but not UUID or libxml2. If you can narrow this down, that would be perfect, but if not, maybe add a comment saying "the following dependencies are ported from AliEn-Runtime, not all of them may be required".

ktf commented:
I fully support getting rid of the alien-runtime. it was introduced when the module files where not protected against recursion and it's now completely unnecessary hack. While you are at it, please make sure you do not add the dependencies blindly but only if they are really needed.

adriansev commented:
@ktf @TimoWilken i began to alidistlint all the touched files, and fix as much i can (one commit per file), and i will try to address all your concerns .. but i do not think that i can (i'm able to) address ALL your concerns ... i have no idea what are the strict dependencies of each packages, the focus of this PR was just to get rid of the alien-runtime .. IMHO, optimizations on dependency chain can be done post "alien-runtime era" ..
Also, note that i do not __"introduce"__ anything, it's the same as before, just without alien-runtime..
Thanks you!

adriansev commented:
@TimoWilken @ktf i fixed as much as possible the warnings given by alidistlint in all touched files. each file have it's own commit.
Also, regarding curl presence in python.sh it is in the the master and it was introduced by you Timo (the blame says so) by https://github.com/alisw/alidist/pull/4019

> Finally, I'm not sure that just removing AliEn-Runtime from the modulefiles is a good idea. You should probably add the things that AliEn-Runtime provided to the modulefile (like OpenSSL, zlib, libxml, etc), so that these libraries are available at runtime, like before.

I'm not sure where i'm supposed to add, and what exactly.. any hint? :)

adriansev commented:
Morning'! @ktf @TimoWilken Any idea what is happening with the tests? Thanks a lot!

TimoWilken commented:
> regarding curl presence in python.sh it is in the the master and it was introduced by you Timo (the blame says so) by #4019

Yes, but I only replaced `system-curl` by `curl`. If you go back through the history, you get to commit 12143c3350b6026c4131c55f1c972ac0d68bfb65 that introduces `curl` as a `build_requires`. [`python.sh` at that revision][pysh] fetches pip using curl, but we don't do that anymore. I suspect whoever updated Python to a version that included pip forgot to remove the `curl` dependency.

[pysh]: https://github.com/alisw/alidist/blob/12143c3350b6026c4131c55f1c972ac0d68bfb65/python.sh#L52

> > Finally, I'm not sure that just removing AliEn-Runtime from the modulefiles is a good idea. You should probably add the things that AliEn-Runtime provided to the modulefile (like OpenSSL, zlib, libxml, etc), so that these libraries are available at runtime, like before.
> 
> I'm not sure where i'm supposed to add, and what exactly.. any hint? :)

For example, here you remove AliEn-Runtime from the modulefile of xalienfs:

https://github.com/alisw/alidist/pull/4894/files#diff-56dbc1c8220154bc47c669e9a147334dbdf72601501d6c2e36171ce85ef50d45L71

But the AliEn-Runtime modulefile e.g. sets variables that point to the certificate store in AliEn-CAs, and other things. If xalienfs needs that at runtime, you should load the appropriate modulefiles (e.g. from AliEn-CAs) instead of AliEn-Runtime's modulefile there.

Similarly for xjalienfs:

https://github.com/alisw/alidist/pull/4894/files#diff-968452a664d21b0d4b07ea3e3ecd2741450a8205844a199a0cc0a2d70c99b8e3L57

If the dependencies aren't needed at runtime after all, then remove them completely (or move them to `build_requires` if only required at build time).

> Any idea what is happening with the tests?

The tests will run when you mark this PR as "ready to review", i.e. non-draft. The CI does not run for draft PRs.

adriansev commented:
> > regarding curl presence in python.sh it is in the the master and it was introduced by you Timo (the blame says so) by #4019
> 
> Yes, but I only replaced `system-curl` by `curl`. If you go back through the history, you get to commit [12143c3](https://github.com/alisw/alidist/commit/12143c3350b6026c4131c55f1c972ac0d68bfb65) that introduces `curl` as a `build_requires`. [`python.sh` at that revision](https://github.com/alisw/alidist/blob/12143c3350b6026c4131c55f1c972ac0d68bfb65/python.sh#L52) fetches pip using curl, but we don't do that anymore. I suspect whoever updated Python to a version that included pip forgot to remove the `curl` dependency.

Hi @TimoWilken ! thanks a lot for info, see below my problems :)

so, i removed now the curl dependency from python

> > > Finally, I'm not sure that just removing AliEn-Runtime from the modulefiles is a good idea. You should probably add the things that AliEn-Runtime provided to the modulefile (like OpenSSL, zlib, libxml, etc), so that these libraries are available at runtime, like before.

all this modules where added instead of AliEn-Runtime (i replaced the instances of AliEn-Runtime) with the list of dependencies contained within, so everything should be in place

> > 
> > 
> > I'm not sure where i'm supposed to add, and what exactly.. any hint? :)
> 
> For example, here you remove AliEn-Runtime from the modulefile of xalienfs:
> 
> [#4894 (files)](https://github.com/alisw/alidist/pull/4894/files#diff-56dbc1c8220154bc47c669e9a147334dbdf72601501d6c2e36171ce85ef50d45L71)
> 
> But the AliEn-Runtime modulefile e.g. sets variables that point to the certificate store in AliEn-CAs, and other things. If xalienfs needs that at runtime, you should load the appropriate modulefiles (e.g. from AliEn-CAs) instead of AliEn-Runtime's modulefile there.
> 
> Similarly for xjalienfs:
> 
> [#4894 (files)](https://github.com/alisw/alidist/pull/4894/files#diff-968452a664d21b0d4b07ea3e3ecd2741450a8205844a199a0cc0a2d70c99b8e3L57)
> 
> If the dependencies aren't needed at runtime after all, then remove them completely (or move them to `build_requires` if only required at build time).

oh, i see what you mean! So, on this i'm not sure about 3 things: 
1. the format of the load.. is it ok: `AliEn-CAs/{AliEn-CAs_VERSION}-{AliEn-CAs_REVISION}` ?
i'm not sure what variables are automatically exported
2. where to add this: should i add this running dependency to any and all packages? Or can i add AliEn-CAs as module dependency to `openssl.sh, osx-system-openssl.sh, system-openssl.sh` and this is solved?
even more, does inheritance work? like adding AliEn-CAS to openssl modules allow me to not add it as dependency to any other module because is part of dependency chain of openssl recipes?
3. does any instance package from `requires:` need manual intervention/addition in `module load` line?

> > Any idea what is happening with the tests?
> 
> The tests will run when you mark this PR as "ready to review", i.e. non-draft. The CI does not run for draft PRs.

oh, ok, i will convert to `ready to review` only after you and Giulio give me green light.
Thanks a lot for patience, info and clarifications! :)

adriansev commented:
Guys, let's focus on the goal of this PR: just remove AliEn-runtime, but leaving anything else as it was (so, a completely transparent removal of AliEn-runtime) and leaving anything else (fixes, optimizations) for further PRs (not necessarily mine). If the scope of this became too wide, nothing will be finished and nothing will move forward.
So, please @TimoWilken, with the additional information that for CAs, it does not matter location, only that that location is specified by X509_CERT_DIR (with is already specified by Alien-CAs), could you answer my above points to be able to move forward this PR?
Thanks a lot!!

TimoWilken commented:
Hi @adriansev,

> so, i removed now the curl dependency from python

Thanks!

> all this modules where added instead of AliEn-Runtime (i replaced the instances of AliEn-Runtime) with the list of dependencies contained within, so everything should be in place

You only did this for the `requires` key in the recipe. If the recipe uses `alibuild-generate-module`, this is enough. However, if the modulefile is created manually (like in `xjalienfs.sh`), you must also add the `requires` dependencies to the modulefile manually.

> 1. the format of the load.. is it ok: `AliEn-CAs/{AliEn-CAs_VERSION}-{AliEn-CAs_REVISION}` ? i'm not sure what variables are automatically exported

See the existing entries for the format. For AliEn-CAs, it would be `${ALIEN_CAS_REVISION:+AliEn-CAs/$ALIEN_CAS_VERSION-$ALIEN_CAS_REVISION}`. The rule for the variable names is "all uppercase, with special chars replaced by underscores". You have `<PACKAGE_NAME>_REVISION` and `<PACKAGE_NAME>_VERSION` variables for all packages in `requires` and `build_requires`, unless they are disabled separately.

The surrounding `${ALIEN_CAS_REVISION:+...}` guards against AliEn-CAs being disabled for that particular build (whether by defaults, or by the user on the command line). If disabled, it must not be loaded in the modulefile, as that would fail at runtime.

> 2. where to add this: should i add this running dependency to any and all packages? Or can i add AliEn-CAs as module dependency to `openssl.sh, osx-system-openssl.sh, system-openssl.sh` and this is solved? even more, does inheritance work? like adding AliEn-CAS to openssl modules allow me to not add it as dependency to any other module because is part of dependency chain of openssl recipes?

Modulefiles will recursively load their dependencies, so in theory you could do this, but I don't think the OpenSSL package is the right place to set up AliEn-CAs.

> 3. does any instance package from `requires:` need manual intervention/addition in `module load` line?

Yes, all of them, if and only if the modulefile is created manually.

adriansev commented:
Hi @TimoWilken !

> > all this modules where added instead of AliEn-Runtime (i replaced the instances of AliEn-Runtime) with the list of dependencies contained within, so everything should be in place
> 
> You only did this for the `requires` key in the recipe. If the recipe uses `alibuild-generate-module`, this is enough. However, if the modulefile is created manually (like in `xjalienfs.sh`), you must also add the `requires` dependencies to the modulefile manually.

so, to make a list, the AliEn-Runtime dependencies are/were:
1. GCC-Toolchain: manual modulefile
2. Xcode : missing modulefile (but this is build_requires not requires like in AliEn-Runtime so it should not need modulefile)
3. zlib : manual modulefile
4. libxml2 : manual modulefile
5. OpenSSL : automatic modulefile
6. osx-system-openssl : missing modulefile (given that this is system provided it does not need modulefile)
7. AliEn-CAs : automatic modulefile
8. ApMon-CPP : manual modulefile
9. UUID : missing modulefile (??)

so, i will try to add automatic modulefile generation for each manual module file
case by case and request you confirmation for each one.

> > 1. the format of the load.. is it ok: `AliEn-CAs/{AliEn-CAs_VERSION}-{AliEn-CAs_REVISION}` ? i'm not sure what variables are automatically exported
> 
> See the existing entries for the format. For AliEn-CAs, it would be `${ALIEN_CAS_REVISION:+AliEn-CAs/$ALIEN_CAS_VERSION-$ALIEN_CAS_REVISION}`. The rule for the variable names is "all uppercase, with special chars replaced by underscores". You have `<PACKAGE_NAME>_REVISION` and `<PACKAGE_NAME>_VERSION` variables for all packages in `requires` and `build_requires`, unless they are disabled separately.
> 
> The surrounding `${ALIEN_CAS_REVISION:+...}` guards against AliEn-CAs being disabled for that particular build (whether by defaults, or by the user on the command line). If disabled, it must not be loaded in the modulefile, as that would fail at runtime.

I see! Thanks a lot for info!!

> > 2. where to add this: should i add this running dependency to any and all packages? Or can i add AliEn-CAs as module dependency to `openssl.sh, osx-system-openssl.sh, system-openssl.sh` and this is solved? even more, does inheritance work? like adding AliEn-CAS to openssl modules allow me to not add it as dependency to any other module because is part of dependency chain of openssl recipes?
> 
> Modulefiles will recursively load their dependencies, so in theory you could do this, but I don't think the OpenSSL package is the right place to set up AliEn-CAs.

actually it is, as OpenSSL is actually the piece of software that use those CAs (see https://github.com/openssl/openssl/blob/master/crypto/x509/x509_def.c#L30)

> > 3. does any instance package from `requires:` need manual intervention/addition in `module load` line?
> 
> Yes, all of them, if and only if the modulefile is created manually.

ok, then i will try to convert them all to automatic modulefile generation

Thanks a lot!

adriansev commented:
@TimoWilken I have done the conversion for zlib... could you please take a look and confirm that i did it the right way? Thanks a lot!

TimoWilken commented:
Hi @adriansev, almost! `alibuild-generate-module` generates the `prepend-path LD_LIBRARY_PATH ...` line for you if you give it the `--lib` option, so you shouldn't add that line manually. Also, the original modulefile doesn't add anything to `PATH`, so you don't need the `--bin` option.

adriansev commented:
@TimoWilken thanks a lot! i fixed this! now, on the GCC-Toolchain recipe: i see some heavy customization in there.. is that something that is already covered by alibuild-generate-module (and i can do it by following what i did for zlib) or should i leave it for alibuild experts?

TimoWilken commented:
No problem @adriansev, but why are you changing the GCC-Toolchain recipe? I'd leave that as-is since it doesn't depend on AliEn-Runtime.

I would only recommend changing the modulefiles of things that depended on AliEn-Runtime -- so that would be `libjalieno2.sh`, `libjalienws.sh`, `xalienfs.sh` and `xjalienfs.sh`.

`libpng.sh` I think you can also skip, since it doesn't actually need AliEn-Runtime. I'm not sure why you changed `zlib.sh` as well!

adriansev commented:
@ktf @TimoWilken unless i missed some package, i converted to alibuild-generate-module all involved packages and solved most of the warnings (if missed i will add some more commits).
But in principle this is ready for you to make a hard review before i convert this to `ready to review` and start the huge build process .. let me know if you want me to squash everything together.

adriansev commented:
@ktf @TimoWilken i removed all wrongly introduced _ROOT variables and keep those that were already defined (but removed by me). Let me know if is there anything else to fix before converting to `ready to review`. Thanks a lot!!

TimoWilken commented on xalienfs.sh:
```diff
@@ -3,79 +3,73 @@ version: "%(tag_basename)s"
- - "OpenSSL:(?!osx)"
- - "osx-system-openssl:(osx.*)"
- - AliEn-Runtime
+  - zlib
+  - libxml2
+  - "OpenSSL:(?!osx)"
+  - "osx-system-openssl:(osx.*)"
+  - AliEn-CAs
+  - ApMon-CPP
+  - UUID
+  - XRootD
```
As far as I can tell, xalienfs directly depends on XRootD and OpenSSL. Are the rest really required? If you're not sure, could you remove them and see if that breaks the build?

adriansev replied:
sure, i left UUID as there is a usage of the form of `${UUID_ROOT:+-I$UUID_ROOT/include -L$UUID_ROOT/lib}` in CXXFLAGS

TimoWilken commented on xrootd.sh:
```diff
@@ -119,17 +123,15 @@ if [[ x"$XROOTD_PYTHON" == x"True" ]]; then
-
-cat >> "$MODULEFILE" <<EoF
+cat >> "etc/modulefiles/${PKGNAME}" <<EoF
+setenv XROOTD_ROOT \$::env(BASEDIR)/$PKGNAME/\$version
```
Again, no new `setenv`s please.

adriansev replied:
done

TimoWilken reviewed (commented):
Hi @adriansev, first of all, thanks again for all the effort you've put into this.

I've made some comments below, but overall, this looks very good to me!

Lastly, I'd recommend deleting the `alien-runtime.sh` recipe completely once it's not used any more.

adriansev commented on libjalienws.sh:
```diff
@@ -3,36 +3,29 @@ version: "%(tag_basename)s"
-set LIBJALIENWS_ROOT \$::env(BASEDIR)/$PKGNAME/\$version
-prepend-path LD_LIBRARY_PATH \$LIBJALIENWS_ROOT/lib
+cat >> "etc/modulefiles/${PKGNAME}" <<EoF
+setenv LIBJALIENWS_ROOT \$::env(BASEDIR)/$PKGNAME/\$version
```
done

adriansev commented on libjalieno2.sh:
```diff
@@ -4,39 +4,32 @@ tag: "0.1.4"
-set LIBJALIENO2_ROOT \$::env(BASEDIR)/$PKGNAME/\$version
-prepend-path LD_LIBRARY_PATH \$LIBJALIENO2_ROOT/lib
+cat >> "etc/modulefiles/${PKGNAME}" <<EoF
+setenv LIBJALIENO2_ROOT \$::env(BASEDIR)/$PKGNAME/\$version
```
done

adriansev commented on libpng.sh:
```diff
@@ -1,43 +1,38 @@
+alibuild-generate-module --lib --bin > "etc/modulefiles/${PKGNAME}"
+
+cat >> "etc/modulefiles/${PKGNAME}" <<EoF
+setenv LIBPNG_ROOT \$::env(BASEDIR)/$PKGNAME/\$version
```
done

adriansev commented on root.sh:
```diff
@@ -3,54 +3,66 @@ version: "%(tag_basename)s"
 tag: "v6-26-10-alice5"
 source: https://github.com/alisw/root.git
 requires:
+  - zlib
```
it is used as argument in cmake command, see https://github.com/alisw/alidist/blob/master/root.sh#L154 just under LIBPNG_ROOT
and yes, it is strictly required as ROOT is making use of compression in many subsystems AFAIK

adriansev commented on root.sh:
```diff
@@ -175,44 +181,40 @@ cmake $SOURCEDIR
 
 FEATURES="builtin_pcre mathmore xml ssl opengl minuit2 http
           pythia6 roofit soversion vdt ${CXX17:+cxx17}
-          ${XROOTD_ROOT:+xrootd} ${ALIEN_RUNTIME_ROOT:+monalisa} ${ROOT_HAS_PYTHON:+pyroot}
```
right!! also changed in the cmake arguments

adriansev commented on root.sh:
```diff
@@ -175,44 +181,40 @@ cmake $SOURCEDIR
 
 # Make some CMake files used by other projects relocatable
+# shellcheck disable=SC2046
 sed -i.deleteme -e "s!$BUILDDIR!$INSTALLROOT!g" $(find "$INSTALLROOT" -name '*.cmake') || true
```
done

adriansev commented on xjalienfs.sh:
```diff
@@ -40,24 +41,13 @@ done
+alibuild-generate-module --lib --bin > "etc/modulefiles/${PKGNAME}"
+
+cat >> "etc/modulefiles/${PKGNAME}" <<EoF
+setenv XJALIENFS_ROOT \$::env(BASEDIR)/$PKGNAME/\$version
```
done

adriansev commented on xalienfs.sh:
```diff
@@ -3,79 +3,73 @@ version: "%(tag_basename)s"
+alibuild-generate-module --lib --bin > "etc/modulefiles/${PKGNAME}"
 
+cat >> "etc/modulefiles/${PKGNAME}" <<EoF
+setenv XALIENFS_ROOT \$::env(BASEDIR)/$PKGNAME/\$version
```
done

adriansev commented on python.sh:
```diff
@@ -3,52 +3,64 @@ version: "%(tag_basename)s"
- - "GCC-Toolchain:(?!osx)"
- - libffi
+  - zlib
+  - libxml2
```
no, it does not. done

adriansev commented on python.sh:
```diff
@@ -3,52 +3,64 @@ version: "%(tag_basename)s"
+  - "OpenSSL:(?!osx)"
+  - "osx-system-openssl:(osx.*)"
+  - AliEn-CAs
+  - UUID
```
done

adriansev commented on python.sh:
```diff
@@ -3,52 +3,64 @@ version: "%(tag_basename)s"
+    CPPFLAGS_LIST=$(find "${ext}" -type d -name include -exec echo -I\{\} \; | tr '\n' ' ')
+    CPPFLAGS="${CPPFLAGS_LIST} ${CPPFLAGS}"
 done
-export LDFLAGS=$(echo $LDFLAGS)
-export CPPFLAGS=$(echo $CPPFLAGS)
```
done

adriansev commented:
Hi @TimoWilken Thanks a lot for help! I believe i addressed all your comments, let me know if this is ready for the tests

TimoWilken reviewed (approved):
Thank you @adriansev! This looks good to me now; feel free to mark as ready to review to start the CI.

adriansev commented:
> Thank you @adriansev! This looks good to me now; feel free to mark as ready to review to start the CI.

Thank you @TimoWilken ! Done!

adriansev commented:
@TimoWilken something is really weird, __ALL__ tests stop at the end of Python building without any visible error .. any idea what could be the problem?

adriansev commented:
and on the successful macOS one i see in the log this:
```
++ find /Volumes/build/build/alice-ci-workdir/alidist-o2/sw/INSTALLROOT/bee4760cc7050509eef88fddc6832deedd096114/osx_x86-64/ROOT/v6-26-10-alice5-local8 -name '*.cmake' -execdir sed -i.deleteme -e 's!/Volumes/build/build/alice-ci-workdir/alidist-o2/sw/BUILD/bee4760cc7050509eef88fddc6832deedd096114/ROOT!/Volumes/build/build/alice-ci-workdir/alidist-o2/sw/INSTALLROOT/bee4760cc7050509eef88fddc6832deedd096114/osx_x86-64/ROOT/v6-26-10-alice5-local8!g' '{}' +
sed: ROOTConfig-targets-relwithdebinfo.cmake: No such file or directory
```
which is also weird (if the find found the file, how can it be missing?)

TimoWilken commented on python.sh:
```diff
@@ -3,52 +3,63 @@ version: "%(tag_basename)s"
-  SSL_CERT_FILE: "$(export PATH=$PYTHON_ROOT/bin:$PATH; export LD_LIBRARY_PATH=$PYTHON_ROOT/lib:$LD_LIBRARY_PATH; python -c \"import certifi; print(certifi.where())\")"
+  SSL_CERT_FILE: "$(PATH=$PYTHON_ROOT/bin:$PATH \
+                    LD_LIBRARY_PATH=$PYTHON_ROOT/lib:$LD_LIBRARY_PATH \
+                    python3 -c 'import certifi; print(certifi.where())')"
```
@adriansev for the builders that fail after Python, the error is:

```
2023-03-17@08:05:38:DEBUG:O2Suite:Python:build_O2_alidist-slc7_x86-64: /build.sh: line 218: syntax error near unexpected token `certifi.where'
```

You need to go back to using `\"` for the inner quotes I think... aliBuild's escaping is sometimes nonexistent, so this variable assignment gets turned into the following line in `build.sh`:

```
echo 'export SSL_CERT_FILE="$(PATH=$PYTHON_ROOT/bin:$PATH LD_LIBRARY_PATH=$PYTHON_ROOT/lib:$LD_LIBRARY_PATH python3 -c 'import certifi; print(certifi.where())')"' >> $INSTALLROOT/etc/profile.d/init.sh
```

But really, this is a bug in aliBuild. I'll see if I can fix the escaping in future, but for now, it might be better to go back to `\"`, sorry!

adriansev replied:
done! let's see how it goes :)

adriansev commented:
Hi @TimoWilken there are still some weird errors (what i'm able to tell):
for o2_alidist there is:
```
Info in <TGeoManager::Export>: Exporting ccdb_object FAIR geometry as root file. Optimizations streamed.
[00:09:03][INFO] Init CcdApi with UserAgentID: alientest06-1679098115-hPB3EK, Host: http://alice-ccdb.cern.ch/
[00:09:04][FATAL] Alien Token Check failed - Please get an alien token before running with https CCDB endpoint, or alice-ccdb.cern.ch!
For later analysis we write a core dump to core_dump_25613
```

for o2_alidist_cs8 :
```
      Start  2: testMonitorObject
 1/40 Test  #2: testMonitorObject .......................***Timeout  30.03 sec
Running 4 test cases...
MonitorObject::getName() : No object in this MonitorObject, returning empty string
MonitorObject::getName() : No object in this MonitorObject, returning empty string
2023-03-18 01:39:56.994590     getName : 'asdf'
2023-03-18 01:39:56.994764     GetName : 'asdf'
2023-03-18 01:39:56.995121     title : ''
2023-03-18 01:39:57.558838     ***
2023-03-18 01:39:57.654749     mo : 0x3fe9ca0
2023-03-18 01:39:57.654795     name : asdf
2023-03-18 01:39:57.654820     name : asdf

[1;32;49m*** No errors detected
```

also i noticed the following openssl usage:
```
Found OpenSSL: /usr/lib64/libcrypto.so (found version "1.1.1k")
```
but in the same time i see in LD_LIBRARY_PATH:`/sw/slc8_x86-64/OpenSSL/v1.1.1m-local2/lib`
so, while one can expect for symbols to be compatible, such a mixup can bring unexpected results..

and alidist_O2Suite_o2_macOS-arm just stops at the end of ROOT building

adriansev commented:
@TimoWilken i will close this as it seems impossible to rebase, and i will try to do another try by applying the diff of this branch on a fresh master
